### PR TITLE
MySQL storage backend

### PR DESCRIPTION
**Pull request by pborzenkov**
The pull request adds MySQL storage backend.

### CODE REVIEW

**Comment by pborzenkov** [comment]
Hmm, looks like Tarvis is using MySQL 5.5 which doesn't support fractional DATETIME values. 5.6 is available in 'trusty' CI environment. I'll remove MySQL backend from CI until the switch is made.

File: storage/sql/config.go [hunk 1]
```diff
@@ -59,14 +70,33 @@ func (s *SQLite3) open(logger logrus.FieldLogger) (*conn, error) {
...
+const (
+	// MySQL SSL modes
+	mysqlSSLTrue       = "true"
+	mysqlSSLFalse      = "false"
+	mysqlSSLSkipVerify = "skip-verify"
+	mysqlSSLCustom     = "custom"
+)
+
+// NetworkDB contains options common to SQL databases accessed over network.
+type NetworkDB struct {
```
Comment (rithujohn191) [inline storage/sql/config.go]: Please add individual comments describing each attribute.

File: storage/sql/config.go [hunk 2]
```diff
@@ -145,3 +170,105 @@ func (p *Postgres) open(logger logrus.FieldLogger) (*conn, error) {
...
+func (s *MySQL) open(logger logrus.FieldLogger) (*conn, error) {
+	cfg := mysql.Config{
+		User:   s.User,
+		Passwd: s.Password,
+		DBName: s.Database,
+
+		Timeout: time.Second * time.Duration(s.ConnectionTimeout),
+
+		ParseTime: true,
+		Params: map[string]string{
```
Comment (rithujohn191) [inline storage/sql/config.go]: Please add a comment to explain this

**Comment by pborzenkov** [comment]
@rithujohn191 The problem is that conformance tests require the storage to provide millisecond precision right now. I see several ways to fix this:

- Keep the precision requirement and require MySQL 5.6. This way either CI has to be switched to 'trusty' or MySQL >= 5.6 has to be started in a docker container;
- Relax the precision requirement in the tests and agree to full second precision;
- Extend the Storage interface with the 'TimestampPrecision() time.Duration' method which will be used to report the maximum timestamp precision supported by the storage. And use the method throughout the test code.

Personally, I think the third one is the way to go and it does not require a lot of coding. What do you think?

**Comment by michelaquino** [comment]
Hi guys, what the PR status?

**Comment by pborzenkov** [comment]
Unfortunately, I didn't have time to finish it. I'll try to complete it in the coming days. I'll go with the item 2 from my proposal (relaxing timestamp precision requirements to full second).

**Comment by ericchiang** [comment]
I don't think the dex team has the bandwidth to maintain another storage implementation. If someone wants to step up as a maintainer, we would consider merging, But without someone dedicated to this storage, I don't think this is going to happen today.

**Comment by sagikazarmark** [comment]
Dex supports MySQL now.

**Pull request closed without merging**